### ISSUE TITLE
Store an affiliate's unpaid earnings in the affiliates_wp_affiliates DB table

### ISSUE THREAD

**Issue by sumobi** [AffiliateWP/AffiliateWP#1228]
PRs:
* [x] #1899
* [ ] #1911

The `affiliates_wp_affiliates` database table shows an `earnings` column. This is the total `paid` earnings of an affiliate. It would be useful to also store an affiliate's `unpaid` earnings. 

For example, with our leaderboard add-on we can list affiliates based on their earnings. Customers have requested to show only an affiliate's unpaid earnings, or a combination of both unpaid and paid. This could be possible by looping through every referral based on the status, but it would be much easier/more efficient to store this amount in the affiliates table. 

I'm sure there are also other use-cases where this would be useful. One that comes to mind is replacing the use of `affwp_get_affiliate_unpaid_earnings()` on the affiliate dashboard. We can get the value directly from the affiliate without having to work it out each time.

**Comment by pippinsplugins** [comment]
I agree, though I'd rather do this in a point release after 1.8, perhaps in 1.8.3.

**Comment by DrewAPicture** [comment]
Started a PR, still a WIP: #1899

**Comment by pippinsplugins** [comment]
@DrewAPicture PR is looking really good. Let me know when ready for testing.

**Comment by DrewAPicture** [comment]
The PR is testable. To test the upgrade routine, you'll either need to manually run the `v20_upgrade()` method yourself, or at least set the `affwp_upgrade_needed` setting to true so the notice with the batch processor form displays. I've been running the following using wp shell in wp-cli:

```sh
affiliate_wp()->settings->set( array( 'affwp_upgrade_needed' => true ), $save = true );
```

**Comment by pippinsplugins** [comment]
@DrewAPicture That makes the upgrade notice show up for me but the button doesn't go anywhere and there is no upgrades page in the tools menu.

**Comment by DrewAPicture** [comment]
@pippinsplugins Make sure you're running the latest. I dropped the Upgrade tab approach in favor of just doing the batch process right there in the notice.

**Comment by pippinsplugins** [comment]
There we go. Looks like the upgrade routine worked!

There is a display issue though:
![screen shot 2017-01-10 at 10 34 19 am](https://cloud.githubusercontent.com/assets/1034109/21814886/73baf632-d720-11e6-87d6-261b122167a9.png)

**Comment by pippinsplugins** [comment]
Nevermind, didn't work.

The upgrade routine reported that it was successful but no changes were made to the database structure.

**Comment by DrewAPicture** [comment]
@pippinsplugins Sorry, I assumed that you already ran `affiliate_wp()->affiliates->create_table()` to create the new column. Since we bumped the version so early, re-running the upgrade method is necessary (or at least the parts of the upgrade method that update the schema and set the setting to display the notice).

**Comment by pippinsplugins** [comment]
There we go!

Going to test some more now that I have the upgrade completed.

**Comment by pippinsplugins** [comment]
Seems to be working great. The unpaid amounts are incrementing and decrementing as they should.

**Comment by sumobi** [comment]
So far so good. I did not get an upgrade routine but perhaps I didn't need one with a fresh install? The `unpaid_earnings` column is definitely there, and my version is now set to 2.0 in the options table.

1. Set up fresh site
2. Installed 1.9.8
3. deactivated and activated the issue branch

**Comment by sumobi** [comment]
Ok happy with this, tried everything I could think of, works great.

**Comment by DrewAPicture** [comment]
Merged #1899.

**Comment by pippinsplugins** [comment]
We have a display issue with the upgrade notice still:
![screen shot 2017-01-10 at 6 14 13 pm](https://cloud.githubusercontent.com/assets/1034109/21830211/c40facfa-d760-11e6-8d04-bd61ec7e2f4c.png)

**Comment by DrewAPicture** [comment]
Started a new PR for shaping up this upgrade routine mess: #1911.

**Comment by pippinsplugins** [comment]
@DrewAPicture new PR is looking good. Ready for testing?

**Comment by DrewAPicture** [comment]
@pippinsplugins All that's left is to bring over the decoupling logic from #1910 so the upgrade routine can run globally in the admin. Working on that right now.

**Comment by DrewAPicture** [comment]
@AffiliateWP/core The PR, #1911, is ready for testing.

**Comment by pippinsplugins** [comment]
@DrewAPicture works great! I'm happy with this.

**Comment by DrewAPicture** [comment]
Merged #1911.
